File: 56-merge-intervals/merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), [](const vector<int>& a, const vector<int>& b){
            return a[0] < b[0];
        });

        vector<vector<int>> res;
        int n = intervals.size();

        vector<int> prev = intervals[0];

        for(int i = 1; i < n; i++){
            vector<int> interval = intervals[i];
            if(interval[0] <= prev[1]){
                prev[1] = max(prev[1], interval[1]);
            }else{
                res.push_back(prev);
                prev = interval;
            }
        }

        res.push_back(prev);

        return res;
    }
};

File: 56-merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x:x[0])

        prev = intervals[0]

        res = []

        for i in range(1,len(intervals)):
            interval = intervals[i]

            if prev[1] >= interval[0]:
                prev[1] = max(prev[1], interval[1])
            else:
                res.append(prev)
                prev = interval

        res.append(prev)

        return res

